File: analytics/src/test/java/io/appmetrica/analytics/impl/billing/BillingInfoSenderImplTest.java
package io.appmetrica.analytics.impl.billing;

import io.appmetrica.analytics.billinginterface.internal.Period;
import io.appmetrica.analytics.billinginterface.internal.ProductInfo;
import io.appmetrica.analytics.billinginterface.internal.ProductType;
import io.appmetrica.analytics.coreapi.internal.executors.ICommonExecutor;
import io.appmetrica.analytics.coreutils.internal.executors.SafeRunnable;
import io.appmetrica.analytics.impl.CounterReport;
import io.appmetrica.analytics.impl.component.IReportableComponent;
import io.appmetrica.analytics.testutils.CommonTest;
import java.util.Arrays;
import java.util.List;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.invocation.InvocationOnMock;
import org.mockito.stubbing.Answer;
import org.robolectric.RobolectricTestRunner;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

@RunWith(RobolectricTestRunner.class)
public class BillingInfoSenderImplTest extends CommonTest {

    private IReportableComponent reportableComponent;
    private BillingInfoSenderImpl billingInfoSender;

    @Before
    public void setUp() {
        reportableComponent = mock(IReportableComponent.class);
        ICommonExecutor reportExecutor = mock(ICommonExecutor.class);

        doAnswer(new Answer<Void>() {
            @Override
            public Void answer(InvocationOnMock invocation) throws Throwable {
                ((Runnable) invocation.getArgument(0)).run();
                return null;
            }
        }).when(reportExecutor).execute(any(SafeRunnable.class));
        billingInfoSender = new BillingInfoSenderImpl(reportableComponent, reportExecutor);
    }

    @Test
    public void testSendInfo() {
        final List<ProductInfo> productInfos = getProductInfos();
        billingInfoSender.sendInfo(productInfos);

        verify(reportableComponent, times(productInfos.size())).handleReport(any(CounterReport.class));
    }

    private List<ProductInfo> getProductInfos() {
        return Arrays.asList(
            new ProductInfo(ProductType.INAPP, "sku1", 2, 1, "by", 3, new Period(1, Period.TimeUnit.DAY), 1, new Period(1, Period.TimeUnit.MONTH), "signature1", "token1", 10, true, "{\"productId\":\"sku1\", \"autoRenewing\":\"true\"}"),
            new ProductInfo(ProductType.SUBS, "sku2", 2, 2, "by", 4, new Period(1, Period.TimeUnit.DAY), 1, new Period(1, Period.TimeUnit.MONTH), "signature2", "token2", 11, false, "")
        );
    }
}

File: billing/src/test/java/io/appmetrica/analytics/billing/impl/sender/BillingInfoSenderImplTest.kt
package io.appmetrica.analytics.billing.impl.sender

import io.appmetrica.analytics.billing.impl.Constants
import io.appmetrica.analytics.billinginterface.internal.Period
import io.appmetrica.analytics.billinginterface.internal.ProductInfo
import io.appmetrica.analytics.billinginterface.internal.ProductType
import io.appmetrica.analytics.coreapi.internal.servicecomponents.ServiceComponentModuleReporter
import io.appmetrica.analytics.coreapi.internal.servicecomponents.ServiceModuleCounterReport
import io.appmetrica.analytics.testutils.CommonTest
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.kotlin.any
import org.mockito.kotlin.doAnswer
import org.mockito.kotlin.mock
import org.mockito.kotlin.verify
import org.mockito.kotlin.verifyNoMoreInteractions
import org.robolectric.RobolectricTestRunner
import java.util.concurrent.Executor

@RunWith(RobolectricTestRunner::class)
internal class BillingInfoSenderImplTest : CommonTest() {

    private val firstProductInfo =
        ProductInfo(
            ProductType.INAPP,
            "sku1",
            2,
            1,
            "by",
            3,
            Period(1, Period.TimeUnit.DAY),
            1,
            Period(1, Period.TimeUnit.MONTH),
            "signature1",
            "token1",
            10,
            true,
            "{\"productId\":\"sku1\", \"autoRenewing\":\"true\"}"
        )
    private val firstValueBytes = "first value bytes".toByteArray()

    private val secondProductInfo =
        ProductInfo(
            ProductType.SUBS,
            "sku2",
            2,
            2,
            "by",
            4,
            Period(1, Period.TimeUnit.DAY),
            1,
            Period(1, Period.TimeUnit.MONTH),
            "signature2",
            "token2",
            11,
            false,
            ""
        )
    private val secondValueBytes = "second value bytes".toByteArray()

    private val productInfos = listOf(
        firstProductInfo,
        secondProductInfo
    )

    private val reporter: ServiceComponentModuleReporter = mock()
    private val executor: Executor = mock {
        on { execute(any()) } doAnswer { it.getArgument<Runnable>(0).run() }
    }
    private val converter: ProductInfoConverter = mock {
        on { fromModel(firstProductInfo) } doAnswer { firstValueBytes }
        on { fromModel(secondProductInfo) } doAnswer { secondValueBytes }
    }

    private val billingInfoSender = BillingInfoSenderImpl(
        reporter,
        executor,
        converter
    )

    @Test
    fun sendInfo() {
        billingInfoSender.sendInfo(productInfos)

        verify(reporter).handleReport(
            ServiceModuleCounterReport.Companion.newBuilder()
                .withType(Constants.Events.TYPE)
                .withValueBytes(firstValueBytes)
                .build()
        )
        verify(reporter).handleReport(
            ServiceModuleCounterReport.Companion.newBuilder()
                .withType(Constants.Events.TYPE)
                .withValueBytes(secondValueBytes)
                .build()
        )
        verifyNoMoreInteractions(reporter)
    }
}
